rclcpp/include/rclcpp/parameter_events_subscriber.hpp

```diff
@@ -0,0 +1,183 @@
+// limitations under the License.
+
+#ifndef RCLCPP__PARAMETER_EVENTS_SUBSCRIBER_HPP_
+#define RCLCPP__PARAMETER_EVENTS_SUBSCRIBER_HPP_
+
+#include <string>
+#include <utility>
+#include <unordered_map>
+#include <vector>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/parameter_events_filter.hpp"
+
+namespace rclcpp
+{
+
+class ParameterEventsSubscriber
+{
+public:
+  ParameterEventsSubscriber(
+    rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
+    rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
+    rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
+    const rclcpp::QoS & qos =
+    rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_parameter_events)));
+
+  template<typename NodeT>
+  ParameterEventsSubscriber(
+    NodeT node,
+    const rclcpp::QoS & qos =
+    rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_parameter_events)))
+  : ParameterEventsSubscriber(
+      node->get_node_base_interface(),
+      node->get_node_topics_interface(),
+      node->get_node_logging_interface(),
+      qos)
+  {}
+
+  /// Sets a custom callback for parameter events
+  /**
+   * If no namespace is provided, a subscription will be created for the current namespace
+   * Repeated calls to this function will overwrite the callback
+   * If more than one namespace already has a subscription to its parameter events topic, then the
+   * provided callback will be applied to all of them
+   *
+   * \param[in] callback Function callback to be evaluated on event
+   * \param[in] Name of namespace for which a subscription will be created
+   */
+  void set_event_callback(
+    std::function<void(const rcl_interfaces::msg::ParameterEvent::SharedPtr &)> callback,
+    const std::string & node_namespace = "");
+
+  /// adds a custom callback for a specified parameter
+  /**
+   * If a node_name is not provided, defaults to the current node
+   *
+   * \param[in] parameter_name Name of parameter
+   * \param[in] callback Function callback to be evaluated upon parameter event
+   * \param[in] node_name Name of node which hosts the parameter
+   */
+  void register_parameter_callback(
+    const std::string & parameter_name,
+    std::function<void(const rclcpp::Parameter &)> callback,
+    const std::string & node_name = "");
+
+  /// adds a callback to assign the value of a changed parameter to a reference variable
+  /**
+   * If a node_name is not provided, defaults to the current node
+   *
+   * \param[in] parameter_name Name of parameter
+   * \param[in] value Reference to variable receiving update
+   * \param[in] node_name Name of node which hosts the parameter
+   */
+  template<typename ParameterT>
+  void register_parameter_update(
+    const std::string & parameter_name, ParameterT & value, const std::string & node_name = "")
+  {
+    auto callback =
+      [&value, this](const rclcpp::Parameter & param) {
```

Comment:
Yes, a user could call `get_parameter` in a custom callback set via `register_parameter_callback`, but that would only apply for parameters existing on the node itself, not for remote nodes for which it has subscribed parameters. To access those parameters, it would need to create a `AsyncParameterClient` or `SyncParametersClient` to that node and call `get_parameter` to make a service call. I was purposely trying to move away from requiring the clients for this use-case. That could be all left to the user, I suppose, but I thought it convenient to provide the utility to update variables automatically.

Concerning thread-safety, I'd like to look into making these methods thread safe using mutex locks to address the concerns you've mentioned.

Reply:
It's true that this adds the benefit of getting a parameter update from an external node without using the service.

The problem is that the reader of the variable that's being updated can be "external" to the class, so I don't see a way of using a mutex lock in a clear way here.
It may be clearer if the user directly writes its callback, using a `mutex` as needed.

Reply:
You can have a callback that gives the user the parameter (extracted from the event rather than using get_parameter), and the user can decide how to get it out of the callback safely. I think this is the easiest thing for us to do.

Or we could have a proxy class where the user needs to access it via a method, like `get()` which can acquire a lock before getting the variable out as a copy (not a reference), protecting it from asynchronous updates.